parsers/GB_NIR.py

```diff
@@ -0,0 +1,166 @@
+from datetime import datetime, timedelta
+from logging import Logger, getLogger
+from typing import Optional
+
+from dateutil import tz
+from requests import ConnectionError, Response, Session
+
+from parsers.lib.config import refetch_frequency
+
+TZ = "Europe/Dublin"
+DATA_URL = "https://www.smartgriddashboard.com/DashboardService.svc/data"
+
+
+def _get_previous_quarter(time: datetime):
+    last_quarter_minute = 15 * (time.minute // 15)
+    return time.replace(minute=last_quarter_minute)
+
+
+def _parse_effective_time(time_str: str):
+    naive_dt = datetime.strptime(time_str, "%d-%b-%Y %H:%M:%S")
+    return naive_dt.replace(tzinfo=tz.gettz(name=TZ))
+
+
+def _fetch_json_data(
+    region: str,
+    dataset: str,
+    session: Session = Session(),
+    target_datetime: Optional[datetime] = None,
+) -> dict:
+    try:
+        if not target_datetime:
+            # get the previous datapoint as it's usually available
+            target_datetime = datetime.now(tz.gettz(name=TZ)) - timedelta(minutes=15)
+
+        previous_quarter = _get_previous_quarter(target_datetime)
+
+        date = previous_quarter.strftime("%d-%b-%Y+%H:%M")
+
+        params = (
+            "?area="
+            + dataset
+            + "&region="
+            + region
+            + "&datefrom="
+            + date
+            + "&dateto="
+            + date
+        )
+
+        response: Response = session.get(DATA_URL + params)
+
+        return response.json()["Rows"][0]
+    except ConnectionError as e:
+        print("Failed to connect to SmartGrid Dashboard:" + e)
+        raise ConnectionError
+
+
+@refetch_frequency(timedelta(minutes=15))
+def fetch_consumption(
+    zone_key: str = "GB-NIR",
+    session: Session = Session(),
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> dict:
+    try:
+        demand = _fetch_json_data("NI", "demandactual", session, target_datetime)
+        demand_mw = float(demand["Value"])
+
+        return {
+            "zoneKey": zone_key,
+            "datetime": _parse_effective_time(demand["EffectiveTime"]),
+            "consumption": demand_mw,
+            "source": "smartgriddashboard.com",
+        }
```

Comment:
The API returns multiple datapoints, it would be great if we parsed them all and returned a list of datapoints, that way the parser is more robust in case of failure or if we are unable to fetch for one time period.

Reply:
The parser retrieves now a day's worth of data and returns the valid datapoints with number values as a list of dicts. This might need to be changed in future if we want to build a fancier breakdown of thermal generation, as it would require using the real-time and previously retrieved fuelmix data (as these are 24h aggregate values in MWh).

Reply:
Nice!

I'll do another review as soon as I have the time but I don't think there should be any issues with it now.

I'll double check with the team about the data but in either case it would be nice to get this merged for the exchanges alone.